signac/synced_collections/data_types/synced_collection.py

```diff
@@ -188,7 +188,9 @@ def __init__(self, parent=None, *args, **kwargs):
             self._load_and_save = self._LoadSaveType(self)
 
         if self._supports_threading:
-            self._locks[self._lock_id] = RLock()
+            with self._cls_lock:
+                if self._lock_id not in self._locks:
+                    self._locks[self._lock_id] = RLock()
```

Comment:
@vyasr I'm wondering whether `defaultdict(RLock)`  might be thread-safe?
https://stackoverflow.com/a/17682555

Reply:
In my opinion it is usually inadvisable to rely on the thread-safety of Python built-ins. Their thread-safety is an implementation detail, and it's also often quite difficult to predict exactly how a given line of Python code will translate to Python bytecode (and Python is only thread-safe at the level of a single bytecode instruction). For example, one thread executing line 193 could create an RLock, but then be preempted by another thread before actually assigning it. We're better off adding a lock to guarantee safety, especially since it's a relatively cheap operation that only occurs on construction.

Reply:
Good point!